### PR TITLE
ENH: Refactor BSplineTransform

### PR DESCRIPTION
**Pull request by blowekamp**
Improve consistency and efficiency of BSplineTransform's
parameters. The set of TransformDomain IVAR have been remove, and the
accessors have been replaced which directly compute the value from the
fixed parameters. The TransformDomain set methods, now call a single
method which computes the fixed parameters from all the transform
domain parameters. The use for the fixed parameters for the
authoritative value of TransformDomain values follows from the fixed
parameters being the serialized values.

Additionally the extra allocation of the CoefficientImage buffers has
been removed, and replaced with the allocation of InternalParameters
and setting of their buffers.

### CODE REVIEW

File: Modules/Core/Transform/test/itkBSplineTransformGTest.cxx [hunk 1]
```diff
@@ -56,9 +57,9 @@ void bspline_eq( const itk::BSplineTransform<TParametersValueType, NDimensions,
     typename ImageType::Pointer coeffImage1 = bspline1->GetCoefficientImages()[i];
     typename ImageType::Pointer coeffImage2 = bspline2->GetCoefficientImages()[i];
 
-    EXPECT_EQ( coeffImage1->GetOrigin(), coeffImage2->GetOrigin() ) << description;
+    EXPECT_VECTOR_NEAR( coeffImage1->GetOrigin(), coeffImage2->GetOrigin(), tolerance ) << description;
     EXPECT_EQ( coeffImage1->GetDirection(), coeffImage2->GetDirection() ) << description;
```
Comment (dzenanz) [inline Modules/Core/Transform/test/itkBSplineTransformGTest.cxx]: No tolerance for direction?

File: Modules/Core/Transform/include/itkBSplineTransform.hxx [hunk 2]
```diff
@@ -154,50 +171,65 @@ void
...
+  for( unsigned int di = 0; di < NDimensions; di++ )
     {
-    this->m_TransformDomainOrigin[j] = this->m_CoefficientImages[0]->GetOrigin()[j] + origin[j];
+    for( unsigned int dj = 0; dj < NDimensions; dj++ )
+      {
+      this->m_FixedParameters[3 * NDimensions + ( di * NDimensions + dj )] =
+        static_cast<FixedParametersValueType>( direction[di][dj] );
+      }
     }
+  std::cout << "FIXED: " << this->m_FixedParameters << std::endl;
```
Comment (dzenanz) [inline Modules/Core/Transform/include/itkBSplineTransform.hxx]: Remnant of debugging?

**Comment by blowekamp** [comment]
This does a good job of refactoring the FixedParameters and removal of the TransformDomain IVARs. There are some style things to clean up still.

Then there is an issue with the m_InternalParametersBuffer and m_Parameters appearing to be duplicates can causing an extra copy of parameters ( this could cause an extra copy for each evaluation 😨 ).

File: Modules/Core/Transform/test/itkBSplineTransformGTest.cxx (hunk 1 above)
Comment (blowekamp) [reply Modules/Core/Transform/test/itkBSplineTransformGTest.cxx]: There are two reasons for this:
 1. The directions are directly stored in the fixed params ( actually mesh size is too and could be EXPECT_EQ too).
2. The `EXPECT_VECTOR_NEAR` does not correctly work on the `Matrix` type.

**Pull request merged**